devito/finite_differences/derivative.py

```diff
@@ -497,3 +580,141 @@ def _eval_fd(self, expr, **kwargs):
             res = res.xreplace(e)
 
         return res
+
+    def _eval_expand_nest(self, **hints):
+        """
+        Expands nested derivatives
+        `Derivative(Derivative(f(x), (x, b)), (x, a))
+            --> Derivative(f(x), (x, a+b))`
+        `Derivative(Derivative(f(x), (y, b)), (x, a))
+            --> Derivative(f(x), (x, a), (y, b))`
+        Note that this is not always a valid expansion depending on the kwargs
+        used to construct the derivative.
+        """
+        if not isinstance(self.expr, self.__class__):
+            return self
+
+        # This is necessary as tools.abc.Reconstructable._rebuild will copy
+        # all kwargs from the self object. Need to enssure that the nest is not
+        # actually expanded if derivatives are incompatible.
+        # The nested derivative is evaluated by:
+        # 1. Chaining together the variables with which to differentiate wrt
+        new_expr = self.expr.args[0]
+        new_dims = [
+            (d, ii)
+            for d, ii in zip(
+                chain(self.dims, self.expr.dims),
+                chain(self.deriv_order, self.expr.deriv_order),
+                strict=True
+            )
+        ]
+
+        # 2. Count the number of derivatives to take wrt each variable as well as
+        # the finite difference order to use by iterating over the chained lists of
+        # variables.
+        new_deriv_order = tuple(chain(self.deriv_order, self.expr.deriv_order))
+        new_fd_order = tuple(chain(self.fd_order, self.expr.fd_order))
+        dcounter = defaultdict(int)
+        fcounter = defaultdict(int)
+        for d, do, fo in zip(new_dims, new_deriv_order, new_fd_order, strict=True):
+            if isinstance(d, Iterable):
+                dcounter[d[0]] += d[1]
+                fcounter[d[0]] += fo
+            else:
+                dcounter[d] += do
+                fcounter[d] += fo
+
+        # 3. Validate that the number of derivatives taken and the `fd_order` are
+        # smaller than or equal to the corresponding space or time order that the
+        # function supports.
+        for (d, do), (_, fo) in zip(dcounter.items(), fcounter.items(), strict=True):
+            if getattr(d, 'is_Time', False):
+                dim_name = 'time'
+                order = self.expr.time_order
+            else:
+                dim_name = 'space'
+                order = self.expr.space_order
+            # The `fd_order` may need to be reduced to construct the nested derivative
+            # in this case we only emit a warning
+            if fo > order:
+                if do > order:
+                    raise ValueError(
+                        f'Nested {do}-derivative constructed which is bigger '
+                        f'than the {dim_name}_order={order}'
+                    )
+                else:
+                    warn(
```

Comment:
why would the order be increased here? If the user specified a smaller  fd_order it should be kept.

Reply:
If the user is taking the `Derivative(..., fd_order=2)` of a `Derivative(..., fd_order=2)` what should the result be? I think you want an `fd_order=4` derivative, but maybe not?

Reply:
I think this is resolved, open again if you aren't convinced